include/nbl/system/demote_promote_writer_readers_lock.h

```diff
@@ -0,0 +1,217 @@
+#ifndef __NBL_DEMOTE_PROMOTE_WRITER_READERS_LOCK_H_INCLUDED__
+#define __NBL_DEMOTE_PROMOTE_WRITER_READERS_LOCK_H_INCLUDED__
+
+#include <boost/atomic/atomic.hpp>
+#include <thread>
+#include <mutex> // for std::adopt_lock_t
+
+// TODO: Reset spins if have to yield?
+// TODO atomic_unsigned_lock_free? since C++20 (from SReadWriteSpinLock)
```

Comment:
yes, just assert the size

Reply:
but then you need to be a bit more clever with the `StateSemantics` left over bits (if any)

Reply:
actually if you want to use `atomic_unsigned_lock_free` then you need `atomic_unsigned_lock_free::value_type` used throughout instead of `uint32_t` for the conversion op, bitfield, etc.